### PR TITLE
Extended the Dispatcher interface to support writing JSON and retrieving the response Content-Type and created a safe default Dispatcher.

### LINKED ISSUE #60: The dispatcher can write a response without providing a Content-Type header.
Currently the `safehttp.Dispatcher` can write a response without providing the `Content-Type` header. This could potentially lead to security issues. We therefore want to ensure that a `Content-Type` header is always written on each response.

This could be accomplished by creating a wrapper around the `http.ResponseWriter` such that the dispatcher needs to first specify a `Content-Type` before getting access to the responsewriter.

### PR DESCRIPTION
**Pull request by maramihali**
Fixes #60, #133

Extended the Dispatcher interface to support writing JSON safe responses and retrieving the Content-Type of safe response and added a . This way the ResponseWriter can query the Dispatcher to retrieve and set the Content-Type of a safe response before writing the response. Implemented and write tests for a default safe dispatcher.

### CODE REVIEW

File: safehttp/default_dispatcher.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+	"github.com/google/safehtml"
+	"github.com/google/safehtml/template"
+	"net/http"
+)
+
+// DefaultDispatcher is responsible of writing the body of a response to an
+// Response Writer if it's of a safe type.
+type DefaultDispatcher struct{}
+
+// ContentType returns the Content-Type of a safe respons if it's of a safe-type
```
Comment (mattiasgrenfeldt) [inline safehttp/default_dispatcher.go:29]: `response` with an `e`

File: safehttp/default_dispatcher.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+// and a non-nil error otherwise.
```
Comment (mattiasgrenfeldt) [inline safehttp/default_dispatcher.go:30]: `and an error otherwise`

File: safehttp/mux_test.go [hunk 3]
```diff
@@ -482,3 +500,12 @@ func TestMuxHandlerReturnsNotWritten(t *testing.T) {
...
 	}
 }
+
+func TestMuxNilDispatcher(t *testing.T) {
+	defer func() {
+		if r := recover(); r == nil {
+			t.Errorf(`mux.NewServeMux(nil): expected panic`)
+		}
+	}()
+	_ = safehttp.NewServeMux(nil)
```
Comment (mattiasgrenfeldt) [inline safehttp/mux_test.go:510]: You can remove the `_ =`

File: safehttp/response_writer.go [hunk 4]
```diff
@@ -64,21 +66,68 @@ func NotWritten() Result {
...
+// TODO: replace panics with proper error handling when getting the response
+// Content-Type or writing the response fails.
 func (w *ResponseWriter) Write(resp Response) Result {
-	w.markWritten()
+	ct, err := w.d.ContentType(resp)
+	if err != nil {
+		panic(err)
+	}
+	w.rw.WriteHeader(int(StatusOK))
+	w.rw.Header().Set("Content-Type", ct)
```
Comment (mattiasgrenfeldt) [inline safehttp/response_writer.go:82]: You need to set the ct header before calling `WriteHeader`. The headers of the response are written when you call `WriteHeader`, so they must be modified first.

The underlying responsewriter will write a 200 OK response if `WriteHeader` was never called, so you can remove the call to `WriteHeader`. And if there is a error during `w.d.Write` which results in a panic then something else will try to write to the responsewriter and in that case it would be bad if the response code is already set.

So remove the call to `WriteHeader` here, in `WriteJSON` and in `WriteTemplate`.

File: safehttp/response_writer_test.go [hunk 5]
```diff
@@ -66,6 +67,18 @@ func TestResponseWriterWriteTwicePanic(t *testing.T) {
...
+			name: "Call WriteJSON twice",
+			write: func(w *safehttp.ResponseWriter) {
+				data := struct{ Field string }{Field: "myField"}
+				jsonObj, err := json.Marshal(data)
+				if err != nil {
+					t.Fatalf("invalid json object")
+				}
+				w.WriteJSON(jsonObj)
+				w.WriteJSON(jsonObj)
+			},
```
Comment (mattiasgrenfeldt) [inline safehttp/response_writer_test.go:80]: When you use `WriteJSON` you send the actual struct, right? The dispatcher will marshal it for you so that isn't needed here. Now you are sending just a string.

File: safehttp/response_writer.go [hunk 6]
```diff
@@ -52,7 +52,9 @@ func (w *ResponseWriter) Interceptor(key string) Interceptor {
 }
 
 // Result TODO
-type Result struct{}
+type Result struct {
+	err error
+}
```
Comment (mattiasgrenfeldt) [inline safehttp/response_writer.go:57]: Is this field used anywhere? If it isn't, remove it for now.

File: safehttp/default_dispatcher_test.go [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+		write func(w http.ResponseWriter) error
+	}{
+		{
+			name: "Unsafe HTML Response",
+			write: func(w http.ResponseWriter) error {
+				d := &safehttp.DefaultDispatcher{}
+				return d.Write(w, "<h1>Hello World!</h1>")
+			},
+		},
+
```
Comment (mattiasgrenfeldt) [inline safehttp/default_dispatcher_test.go:94]: Remove empty line

**Review by mattiasgrenfeldt — CHANGES_REQUESTED** [review]

File: safehttp/default_dispatcher.go (hunk 1 above)
Comment (kele) [reply safehttp/default_dispatcher.go:29]: FYI: @mattiasgrenfeldt check out the "add suggestion" option when making comments (the +- sign in the comment toolbar). This allows the PR author to apply the suggestions from GitHub's web UI with one click.

File: safehttp/default_dispatcher.go [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+	_, err := rw.Write([]byte(x.String()))
+	return err
+
+}
+
+// WriteJSON serialises and writes a JSON to the Response Writer. If the
+// provided response is not valid JSON or writing the response fails, the method
+// will return an error.
+func (DefaultDispatcher) WriteJSON(rw http.ResponseWriter, resp JSONResponse) error {
+	obj, err := json.Marshal(resp.Data)
```
Comment (kele) [inline safehttp/default_dispatcher.go:59]: Add a `TODO` explaining that this is **not** safe yet, and XSSI protection needs to be added (where?).

File: safehttp/default_dispatcher.go [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+// ContentType returns the Content-Type of a safe respons if it's of a safe-type
+// and a non-nil error otherwise.
+func (DefaultDispatcher) ContentType(resp Response) (string, error) {
+	switch resp.(type) {
+	case safehtml.HTML, *template.Template:
+		return "text/html; charset=utf-8", nil
+	case JSONResponse:
+		return "application/json; charset=utf-8", nil
+	default:
+		return "", errors.New("not a safe response")
```
Comment (kele) [inline safehttp/default_dispatcher.go:38]: More descriptive errors please (e.g. include type).

File: safehttp/default_dispatcher.go [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+	}
+}
+
+// Write writes the response to the Response Writer if it's safe HTML. It
+// returns a non-nil error if the response is not safe HTML or if writing the
+// response fails.
+func (DefaultDispatcher) Write(rw http.ResponseWriter, resp Response) error {
+	x, ok := resp.(safehtml.HTML)
+	if !ok {
+		return errors.New("not a safe response type")
```
Comment (kele) [inline safehttp/default_dispatcher.go:48]: More descriptive errors please (e.g. include type).

File: safehttp/default_dispatcher.go [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+
+}
+
+// WriteJSON serialises and writes a JSON to the Response Writer. If the
+// provided response is not valid JSON or writing the response fails, the method
+// will return an error.
+func (DefaultDispatcher) WriteJSON(rw http.ResponseWriter, resp JSONResponse) error {
+	obj, err := json.Marshal(resp.Data)
+	if err != nil {
+		return errors.New("invalid json")
```
Comment (kele) [inline safehttp/default_dispatcher.go:61]: More descriptive errors please (e.g. include `err`, or simply return it).

File: safehttp/default_dispatcher.go [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+	}
+	_, err = rw.Write(obj)
+	return err
```
Comment (kele) [inline safehttp/default_dispatcher.go:64]: Use `json.Encoder` instead, i.e.: `json.NewEncoder(rw).Encode(obj)`. Now the marshalling to an object is unnecessary, as it can be done on the fly by the `Encoder`.

File: safehttp/default_dispatcher.go [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+// ExecuteTemplate applies a parsed template to the provided data object if the
+// template is a safe HTML template, writing the output to the ResponseWriter.
+//
+// If an error occurs executing the template or writing its output,
+// execution stops, but partial results may already have been written to
+// the Response Writer.
+func (DefaultDispatcher) ExecuteTemplate(rw http.ResponseWriter, t Template, data interface{}) error {
+	x, ok := t.(*template.Template)
+	if !ok {
+		return errors.New("not a safe response type")
```
Comment (kele) [inline safehttp/default_dispatcher.go:76]: More descriptive errors please (e.g. include type).

File: safehttp/default_dispatcher.go [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+	}
+	err := x.Execute(rw, data)
+	return err
```
Comment (kele) [inline safehttp/default_dispatcher.go:79]:
```suggestion
	return x.Execute(rw, data)
```

File: safehttp/response_writer.go (hunk 4 above)
Comment (kele) [reply safehttp/response_writer.go:82]: 1. I agree that the headers need to be set before doing `WriteHeader`.
2. I'd still explicitly call `WriteHeader(200)` here, though.

File: safehttp/response_writer.go [hunk 15]
```diff
@@ -64,21 +66,68 @@ func NotWritten() Result {
...
+	if err != nil {
+		panic(err)
+	}
+	w.rw.WriteHeader(int(StatusOK))
+	w.rw.Header().Set("Content-Type", ct)
 	if err := w.d.Write(w.rw, resp); err != nil {
-		panic("error")
+		panic(err)
 	}
+	w.markWritten()
```
Comment (kele) [inline safehttp/response_writer.go:86]: Why this change?

File: tests/integration/header/header_test.go [hunk 16]
```diff
@@ -29,6 +32,17 @@ import (
...
+	switch resp.(type) {
+	case safehtml.HTML, *template.Template:
+		return "text/html; charset=utf-8", nil
+	case safehttp.JSONResponse:
+		return "application/json; charset=utf-8", nil
+	default:
+		return "", errors.New("not a safe response")
+	}
+}
+
```
Comment (kele) [inline tests/integration/header/header_test.go:45]: Please file an issue to carve out the `testDispatcher`, as it feels it should be the same for most of the tests.

**Review by kele — CHANGES_REQUESTED** [review]

**Comment by maramihali** [comment]
I realised I forgot to add the integration tests that test a `ServeMux` created with a `DefaultDispatcher`, I amended the last commit. Sorry about this!

File: safehttp/response_writer.go (hunk 15 above)
Comment (maramihali) [reply safehttp/response_writer.go:73]: We decided in the simplifying session to only call `markWritten` after a response has been written and this change made its way in this PR as I was refactoring things.

File: safehttp/response_writer.go (hunk 4 above)
Comment (mattiasgrenfeldt) [reply safehttp/response_writer.go]: Imagine that `w.d.Write` panics or returns an error, in which case `ResponseWriter.Write` will panic. This panic will be recovered by the framework (ServeMux) and a 500 Internal Server Error response will be written to the `ResponseWriter`. But the status code and the headers have already been written, so things will break.

We need to remove the call to `WriteHeader`.

Comment (maramihali) [reply safehttp/response_writer.go]: Right, we discussed this offline yesterday. We need a better way of handling errors instead of the panics. Right now, if the Dispatcher errs after writing half of the response, the status code and content type had already been set and there's not much we can do as we cannot reset it, right? An option is to buffer the response and only send it after we make sure everything works as intended, but IIRC that's a feature that we don't plan to introduce in V0.

Also, BTW, these are problems that don't come up in tests because we use the `responseRecorder`. This is not ideal IMO as it's easy to introduce bugs (e.g. my bad ordering of statements) and tests won't fail. Should we do something about it?

File: safehttp/default_dispatcher.go [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+
+// Write writes the response to the Response Writer if it's safe HTML. It
+// returns a non-nil error if the response is not safe HTML or if writing the
+// response fails.
+func (DefaultDispatcher) Write(rw http.ResponseWriter, resp Response) error {
+	x, ok := resp.(safehtml.HTML)
+	if !ok {
+		return fmt.Errorf("%T is not a safe response type and it cannot be written", resp)
+	}
+	_, err := rw.Write([]byte(x.String()))
```
Comment (kele) [inline safehttp/default_dispatcher.go:50]: Use `io.WriteString`. If `http.ResponseWriter` implements the `StringWriter` interface, this might save allocating a byte slice (and it seems it [does](https://golang.org/src/net/http/serve_test.go#L4015)).

File: safehttp/default_dispatcher.go [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/google/safehtml"
+	"github.com/google/safehtml/template"
+	"net/http"
+)
+
+// DefaultDispatcher is responsible of writing the body of a response to an
+// Response Writer if it's of a safe type.
```
Comment (kele) [inline safehttp/default_dispatcher.go:26]:
```suggestion
// DefaultDispatcher is responsible for writing safe responses.
```

File: safehttp/response_writer.go (hunk 15 above)
Comment (kele) [reply safehttp/response_writer.go:73]: Please leave out this change for a separate PR.

File: safehttp/response_writer.go [hunk 19]
```diff
@@ -122,5 +169,7 @@ func (w *ResponseWriter) SetCookie(c *Cookie) error {
 // Dispatcher TODO
 type Dispatcher interface {
 	Write(rw http.ResponseWriter, resp Response) error
+	WriteJSON(rw http.ResponseWriter, resp JSONResponse) error
 	ExecuteTemplate(rw http.ResponseWriter, t Template, data interface{}) error
+	ContentType(resp Response) (string, error)
```
Comment (kele) [inline safehttp/response_writer.go:174]: Start adding comments to the methods. You don't have to add to the ones that existed before your PR, we can do that later.

File: safehttp/response_writer_test.go [hunk 20]
```diff
@@ -66,6 +66,14 @@ func TestResponseWriterWriteTwicePanic(t *testing.T) {
 				w.Write(safehtml.HTMLEscaped("<h1>Escaped, so not really a heading</h1>"))
 			},
 		},
+		{
+			name: "Call WriteJSON twice",
+			write: func(w *safehttp.ResponseWriter) {
+				obj := struct{ Field string }{Field: "myField"}
+				w.WriteJSON(obj)
+				w.WriteJSON(obj)
+			},
```
Comment (kele) [inline safehttp/response_writer_test.go:75]: Add a tests that tries to `w.Write` and then `w.WriteJSON`.

File: safehttp/safehttptest/recorder.go [hunk 21]
```diff
@@ -26,6 +28,17 @@ import (
 
 type testDispatcher struct{}
```
Comment (kele) [inline safehttp/safehttptest/recorder.go:29]: Does this mean you could use the `DefaultDispatcher` here now? Possibly in other tests too?

This can be done in a follow-up PR.

**Review by kele — CHANGES_REQUESTED** [review]

File: safehttp/safehttptest/recorder.go (hunk 21 above)
Comment (maramihali) [reply safehttp/safehttptest/recorder.go:29]: Yep, I think so, and it would definitely make tests cleaner. We could use the `DefaultDispatcher` in tests by default (too many defaults in one sentence) and only create a `testDispatcher` for analysing some custom behaviour. 

But yeah, should be done in a clean-up PR and not in this one, I think.

File: safehttp/response_writer.go (hunk 15 above)
Comment (maramihali) [reply safehttp/response_writer.go:73]: The big problem with this is that it introduces a bug, making the tests  panic. So, if we put `markWritten` at the beginning that will set `w.written` to `true`. Then if either `d.ContentType(...)` or `d.WriteSomething(...)` returns an error, we will panic and that will be recovered by the `ServeMux`. The `ServeMux` will then attempt to call`WriteError` with a 500, which attempts to call `markWritten` again, which will panic because `w.written` was already set to true which ends up in a recursive thing, I think :/

Comment (maramihali) [reply safehttp/response_writer.go:73]: Therefore, I'd suggest leaving it as it is in this PR and then fixing it properly in a follow-up PR (the alternative is `t.Skip()` the tests in `integration/mux/mux_test.go` that test the behaviour of passing unsafe responses until we fix it properly)

File: safehttp/response_writer.go (hunk 4 above)
Comment (mattiasgrenfeldt) [reply safehttp/response_writer.go]: I am not sure what to do here.

File: safehttp/default_dispatcher.go [hunk 22]
```diff
@@ -0,0 +1,77 @@
...
+	return err
+
+}
+
+// WriteJSON serialises and writes a JSON to the underlying http.ResponseWriter.
+// If the provided response is not valid JSON or writing the response fails, the
+// method will return an error.
+func (DefaultDispatcher) WriteJSON(rw http.ResponseWriter, resp JSONResponse) error {
+	// TODO: XSSI protection needs to be added as this is not a safe response
+	// type yet
```
[the remark this answers is not in the document]
Comment (maramihali) [reply safehttp/default_dispatcher.go:60]: Oh, how is that preventing XSSI?

File: safehttp/default_dispatcher.go [hunk 23]
```diff
@@ -0,0 +1,77 @@
...
+	"io"
+	"net/http"
+)
+
+// DefaultDispatcher is responsible for writing safe responses.
+type DefaultDispatcher struct{}
+
+// ContentType returns the Content-Type of a safe response if it's of a
+// safe-type and an error otherwise.
+func (DefaultDispatcher) ContentType(resp Response) (string, error) {
```
[the remark this answers is not in the document]
Comment (maramihali) [reply safehttp/default_dispatcher.go:31]: Wasn't sure whether that is okay, will add :)

File: safehttp/response_writer.go [hunk 24]
```diff
@@ -53,20 +53,67 @@ func NotWritten() Result {
 	return Result{}
 }
 
-// Write TODO
+// Write dispatches the response to the Dispatcher, setting the Content-Type to
+// text/html; charset=utf-8 and response status to 200 OK if resp is a safe
```
Comment (mattiasgrenfeldt) [inline safehttp/response_writer.go:57]: This sets the `Content-Type` to whatever the dispatcher tells it to. Not necessarily `text/html`

File: safehttp/response_writer.go [hunk 25]
```diff
@@ -53,20 +53,67 @@ func NotWritten() Result {
...
+// TODO: replace panics with proper error handling when getting the response
+// Content-Type or writing the response fails.
+func (w *ResponseWriter) WriteJSON(data interface{}) Result {
+	w.markWritten()
+	resp := JSONResponse{Data: data}
+	ct, err := w.d.ContentType(resp)
+	if err != nil {
+		panic(err)
+	}
+	w.rw.Header().Set("Content-Type", ct)
```
Comment (mattiasgrenfeldt) [inline safehttp/response_writer.go:91]: Shouldn't this method always set the Content-Type to `application/json`? Then it doesn't need to interrogate the Dispatcher

File: safehttp/response_writer.go [hunk 26]
```diff
@@ -53,20 +53,67 @@ func NotWritten() Result {
...
+}
+
+// WriteTemplate dispatches a parsed template and a data object to the
+// Dispatcher to be executed and written to the underlying Response Writer, in
+// case the template is a safe HTML template. If t is a safe HTML Template, the
+// Content-Type will also be set to text/html; charset=utf-8 and the response
+// status to 200 OK.
+//
+// TODO: replace panics with proper error handling when getting the response
+// Content-Type or writing the response fails.
```
Comment (mattiasgrenfeldt) [inline safehttp/response_writer.go:106]: The `Content-Type` for safe html templates depends on the Dispatcher implementation and should be left out of this comment.

**Review by mattiasgrenfeldt — CHANGES_REQUESTED** [review]

File: safehttp/response_writer.go (hunk 4 above)
Comment (mattiasgrenfeldt) [reply safehttp/response_writer.go]: I agree that we need a better error strategy for the dispatcher, but as you said, that's not V0.

To detect these kinds of errors in tests we need thorough integration tests between the ServeMux, the ResponseWriter and a Dispatcher.

But I still think that the ResponseWriter should not call `WriteHeader`. The Dispatcher should start doing the actual writing to the network connection.

File: safehttp/response_writer_test.go (hunk 20 above)
Comment (mattiasgrenfeldt) [reply safehttp/response_writer_test.go:75]: I opened an issue for testing every combination and ordering of writing methods. #157

File: safehttp/response_writer.go [hunk 27]
```diff
@@ -119,8 +166,39 @@ func (w *ResponseWriter) SetCookie(c *Cookie) error {
...
+	// and writes the output to the http.ResponseWriter.
+	//
+	// This should return an error if the provided Template response is not a
+	// valid template or if an error occurs executing or writing the template.
 	ExecuteTemplate(rw http.ResponseWriter, t Template, data interface{}) error
+
+	// Content-Type returns the Content-Type of the provided response if it is
+	// of a type supported by the Dispatcher and should return an error
+	// otherwise.
+
```
Comment (kele) [inline safehttp/response_writer.go:197]:
```suggestion
	//
```

File: safehttp/response_writer.go [hunk 28]
```diff
@@ -119,8 +166,39 @@ func (w *ResponseWriter) SetCookie(c *Cookie) error {
...
+	// Sending a response to the http.ResponseWriter without properly setting
+	// the is bug-prone and could introduce vulnerabilities. Therefore, this
+	// method should be used to set the Content-Type header before calling any
```
Comment (kele) [inline safehttp/response_writer.go:200]:
```suggestion
	// Sending a response to the http.ResponseWriter without properly setting
	// CT is error-prone and could introduce vulnerabilities. Therefore, this
	// method should be used to set the Content-Type header before calling any
```

File: safehttp/response_writer.go [hunk 29]
```diff
@@ -119,8 +166,39 @@ func (w *ResponseWriter) SetCookie(c *Cookie) error {
...
+	// of the writing methods in the Dispatcher. Writing should not proceed
+	// ContentType returns an error.
```
Comment (kele) [inline safehttp/response_writer.go:202]:
```suggestion
	// if ContentType returns an error.
```

File: safehttp/default_dispatcher.go [hunk 30]
```diff
@@ -0,0 +1,76 @@
...
+	switch resp.(type) {
+	case safehtml.HTML, *template.Template:
+		return "text/html; charset=utf-8", nil
+	case JSONResponse:
+		return "application/json; charset=utf-8", nil
+	default:
+		return "", fmt.Errorf("%T is not a safe response type, a Content-Type cannot be provided", resp)
```
Comment (kele) [inline safehttp/default_dispatcher.go:38]:
```suggestion
		return "", fmt.Errorf("unknown Content-Type for %T", resp)
```

File: safehttp/default_dispatcher.go [hunk 31]
```diff
@@ -0,0 +1,76 @@
...
+	_, err := io.WriteString(rw, x.String())
+	return err
+
+}
+
+// WriteJSON serialises and writes a JSON to the underlying http.ResponseWriter.
+// If the provided response is not valid JSON or writing the response fails, the
+// method will return an error.
+func (DefaultDispatcher) WriteJSON(rw http.ResponseWriter, resp JSONResponse) error {
+	io.WriteString(rw, ")]}',\n")
```
Comment (kele) [inline safehttp/default_dispatcher.go:59]:
```suggestion
	io.WriteString(rw, ")]}',\n") // Break parsing of JavaScript in order to prevent XSSI.
```

File: safehttp/response_writer.go (hunk 25 above)
Comment (kele) [reply safehttp/response_writer.go:91]: I'd keep it for consistency.

File: safehttp/response_writer.go [hunk 32]
```diff
@@ -119,8 +166,39 @@ func (w *ResponseWriter) SetCookie(c *Cookie) error {
 	return w.header.addCookie(c)
 }
 
-// Dispatcher TODO
+// Dispatcher is responsible for writing a response received from the
+// ResponseWriter to the underlying http.ResponseWriter.
+//
+// The implementation of a custom Dispatcher should be thoroughly reviewed by
+// the security team to avoid introducing vulnerabilities. Moreover/
```
Comment (kele) [inline safehttp/response_writer.go:173]: This comment ends abruptly.

File: safehttp/response_writer.go [hunk 33]
```diff
@@ -119,8 +166,39 @@ func (w *ResponseWriter) SetCookie(c *Cookie) error {
...
 	Write(rw http.ResponseWriter, resp Response) error
+
+	// WriteJSON serialises and writes a JSON object to the http.ResponseWriter.
+	//
+	// This should return an error if the provided response is not a valid
+	// JSONResponse  or if writing the object to the http.ResponseWriter fails.
+	WriteJSON(rw http.ResponseWriter, resp JSONResponse) error
+
+	// ExecuteTemplate applies a parsed template to the provided data object
+	// and writes the output to the http.ResponseWriter.
```
Comment (kele) [inline safehttp/response_writer.go:188]:
```suggestion
	// ExecuteTemplate applies a template to the provided data object
	// and writes the output to the http.ResponseWriter.
```

File: safehttp/response_writer.go [hunk 34]
```diff
@@ -119,8 +166,39 @@ func (w *ResponseWriter) SetCookie(c *Cookie) error {
...
+// The implementation of a custom Dispatcher should be thoroughly reviewed by
+// the security team to avoid introducing vulnerabilities. Moreover/
 type Dispatcher interface {
+	// Write writes a Response to the http.ResponseWriter.
+	//
+	// It should return an error if the writing operation fails or if the
+	// provided Response should not be written to the http.ResponseWriter
 	Write(rw http.ResponseWriter, resp Response) error
+
+	// WriteJSON serialises and writes a JSON object to the http.ResponseWriter.
```
Comment (kele) [inline safehttp/response_writer.go:181]:
```suggestion
	// WriteJSON writes a given object as JSON to the http.ResponseWriter.
```

**Review by kele — APPROVED** [review]

**Comment by kele** [comment]
@googlebot I consent.

**Comment by kele** [comment]
@empijei I think you need to do the `@googlebot I consent` thing

File: safehttp/response_writer.go (hunk 15 above)
Comment (maramihali) [reply safehttp/response_writer.go]: Opened #164.

File: safehttp/safehttptest/recorder.go (hunk 21 above)
Comment (maramihali) [reply safehttp/safehttptest/recorder.go:29]: Opened #149

File: safehttp/response_writer.go (hunk 4 above)
Comment (maramihali) [reply safehttp/response_writer.go]: I am not sure I understand the last sentence.

**Comment by empijei** [comment]
@googlebot I consent.
